src/System.Private.CoreLib/shared/System/Text/Rune.cs

```diff
@@ -0,0 +1,724 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Globalization;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    /// <summary>
+    /// Represents a Unicode scalar value ([ U+0000..U+D7FF ], inclusive; or [ U+E000..U+10FFFF ], inclusive).
+    /// </summary>
+    /// <remarks>
+    /// This type's constructors and conversion operators validate the input, so consumers can call the APIs
+    /// assuming that the underlying <see cref="Rune"/> instance is well-formed.
+    /// </remarks>
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public readonly struct Rune : IComparable<Rune>, IEquatable<Rune>
+    {
+        private const byte IS_WHITESPACE_FLAG = 0x80;
+        private const byte IS_LETTER_OR_DIGIT_FLAG = 0x40;
+        private const byte UNICODECATEGORY_MASK = 0x1F;
+
+        // Contains information about the ASCII character range [ U+0000..U+007F ], with:
+        // - 0x80 bit if set means 'is whitespace'
+        // - 0x40 bit if set means 'is letter or digit'
+        // - 0x20 bit is reserved for future use
+        // - bottom 5 bits are the UnicodeCategory of the character
+        private static ReadOnlySpan<byte> AsciiCharInfo => new byte[]
+        {
+            0x0E, 0x0E, 0x0E, 0x0E, 0x0E, 0x0E, 0x0E, 0x0E, 0x0E, 0x8E, 0x8E, 0x8E, 0x8E, 0x8E, 0x0E, 0x0E,
```

Comment:
While ASCII's information probably won't change, it would be nice to see this derived from the same source as the rest of the Unicode categorization, instead of being an independent blob.